src/leafletwidget/widget.ts

```diff
@@ -80,47 +100,46 @@ class MapWidget extends Widget {
     if (!this.isAttached) {
       return;
     }
-    let contentString = this._context.model.toString();
-    if (contentString === this._geojsonString) {
+    let content = this._context.model.toString();
+    let geojson: JSONValue = content ? JSON.parse(content) : content;
+    if (deepEqual(geojson, this._geojson)) {
       return;
     }
 
     // we're attached to the DOM and have new layer content now
     if (this._geojsonLayer) {
       this._map.removeLayer(this._geojsonLayer);
     }
-    this._geojsonString = contentString;
+    this._geojson = geojson;
     this._geojsonLayer = null;
-    if (contentString) {
-      let content = JSON.parse(contentString);
-      this._geojsonLayer = leaflet.geoJson(content, {
+    if (geojson) {
+      this._geojsonLayer = leaflet.geoJson(geojson, {
         pointToLayer: function (feature, latlng) {
             return leaflet.circleMarker(latlng);
         }
       });
       this._map.addLayer(this._geojsonLayer);
-      this._map.fitBounds(this._geojsonLayer.getBounds());
+      this._fitLayerBounds();
     }
   }
 
   /**
    * A message handler invoked on a 'resize' message.
    */
   onResize(msg: ResizeMessage) {
-    // Since we know the size from the resize message, we manually
-    // define getSize() so that it does not have to do a DOM read.
-    this._map.getSize = () => {
-      if (msg.width === -1 || msg.height === -1) {
-        return (this._map as any).prototype.getSize();
-      } else {
-        let size: any = new leaflet.Point(msg.width, msg.height);
-        (this._map as any)._size = size;
-        return size.clone();
-      }
-    };
+    this._width = msg.width;
+    this._height = msg.height;
     this._map.invalidateSize(true);
-    if (this._geojsonLayer) {
+    this._fitLayerBounds();
+  }
+
+  /**
+   * Make the map fit the geojson layer bounds only once when all info is available.
+   */
+  private _fitLayerBounds() {
+    if (this._fitBounds && this._geojsonLayer && this._width && this._height) {
```

Comment:
That's fine - then there is a DOM read. That handles the case when this is not inside a phosphor absolute-position layout.